src/ore/src/metrics.rs

```diff
@@ -403,5 +410,369 @@ pub type ComputedIntGauge = ComputedGenericGauge<AtomicI64>;
 /// A [`ComputedGenericGauge`] for 64-bit unsigned integers.
 pub type ComputedUIntGauge = ComputedGenericGauge<AtomicU64>;
 
+/// Unit of time used for reporting with [`MetricsFutureExt`].
+#[derive(Debug)]
+pub enum TimeUnit {
+    /// Report duration as floating point number of seconds.
+    Seconds,
+    /// Report duration as an integer number of milliseconds.
+    Milliseconds,
+}
+
+/// Exposes combinators that report metrics related to the execution of a [`Future`] to prometheus.
+pub trait MetricsFutureExt<F> {
+    /// Records the amount of time it takes a [`Future`] to complete according to "the clock on
+    /// the wall".
+    ///
+    /// More specifically, it records the instant at which the `Future` was first polled, and the
+    /// instant at which the `Future` completes. Then reports the duration between those two
+    /// instances to the provided histogram.
+    fn record_wall_time(
+        self,
+        histogram: prometheus::HistogramVec,
+        unit: TimeUnit,
+    ) -> WallTimeFuture<F>;
+
+    /// Records the total amount of time for which a [`Future`] was executing.
+    ///
+    /// More specifically, every time the `Future` is polled it records how long that individual
+    /// call took, and maintains a running sum until the `Future` completes. Then we report that
+    /// duration to the provided histogram.
+    fn record_exec_time(
```

Comment:
any early thoughts on when one would want `record_wall_time` vs `record_exec_time`? they both seem pretty interesting, and a large delta between them would also be revealing when understanding system behavior

Reply:
might be useful to capture some recommendation in the doc